pkg/cloud/services/eks/oidc.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	trustPolicyConfigMapName      = "boilerplate-oidc-trust-policy"
+	trustPolicyConfigMapNamespace = metav1.NamespaceDefault
+
+	whitespaceRe = regexp.MustCompile(`(?m)[\t\n]`)
+)
+
+func (s *Service) reconcileOIDCProvider(cluster *eks.Cluster) error {
+	if !s.scope.ControlPlane.Spec.AssociateOIDCProvider || s.scope.ControlPlane.Status.OIDCProvider.ARN != "" {
```

Comment:
We also should have this functionality sit behind the `EKSEnableIAM` feature flag like the rest of the options that create IAM resources. 

If a user chooses to run the managers without this feature flag set to true but they specify a `AssociateOIDCProvider` then it should throw an error.

This can be used on the scope to help with this: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/scope/managedcontrolplane.go#L248

Reply:
👍